src/storage-controller/src/instance.rs

```diff
@@ -0,0 +1,444 @@
+use tracing::{debug, info, warn};
+
+use crate::history::CommandHistory;
+
+/// A controller for a storage instance.
+///
+/// Encapsulates communication with replicas in this instance, and their rehydration.
+///
+/// Note that storage objects (sources and sinks) don't currently support replication (#17418).
+/// An instance can have muliple replicas connected, but only if it has no storage objects
+/// installed. Attempting to install storage objects on multi-replica instances, or attempting to
+/// add more than one replica to instances that have storage objects installed, is illegal and will
+/// lead to panics.
+#[derive(Debug)]
+pub(crate) struct Instance<T> {
+    /// The replicas connected to this storage instance.
+    replicas: BTreeMap<ReplicaId, Replica<T>>,
+    /// The command history, used to replay past commands when introducing new replicas or
+    /// reconnecting to existing replicas.
+    history: CommandHistory<T>,
+    /// The current cluster startup epoch.
+    ///
+    /// The `replica` value of the epoch is increased every time a replica is (re)connected,
+    /// allowing the distinction of different replica incarnations.
+    epoch: ClusterStartupEpoch,
+    /// Metrics tracked for this storage instance.
+    metrics: InstanceMetrics,
+    /// A function that returns the current time.
+    now: NowFn,
+    /// Responses ready for delivery on the next call to [`Instance::recv`].
+    ready_responses: VecDeque<StorageResponse<T>>,
+}
+
+impl<T> Instance<T>
+where
+    T: Timestamp + Lattice,
+    StorageGrpcClient: StorageClient<T>,
+{
+    /// Creates a new [`Instance`].
+    pub fn new(envd_epoch: NonZeroI64, metrics: InstanceMetrics, now: NowFn) -> Self {
+        let history = CommandHistory::new(metrics.for_history());
+        let epoch = ClusterStartupEpoch::new(envd_epoch, 0);
+
+        let mut instance = Self {
+            replicas: Default::default(),
+            history,
+            epoch,
+            metrics,
+            now,
+            ready_responses: Default::default(),
+        };
+
+        instance.send(StorageCommand::CreateTimely {
+            config: TimelyConfig::default(),
+            epoch,
+        });
+
+        instance
+    }
+
+    /// Returns the IDs of all replicas connected to this storage instance.
+    pub fn replica_ids(&self) -> impl Iterator<Item = ReplicaId> + '_ {
+        self.replicas.keys().copied()
+    }
+
+    /// Adds a new replica to this storage instance.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this storage instance has storage objects installed and already has a replica
+    /// connected.
+    pub fn add_replica(&mut self, id: ReplicaId, config: ReplicaConfig) {
+        // Reduce the history to limit the amount of commands sent to the new replica, and to
+        // enable the `objects_installed` assert below.
+        self.history.reduce();
+
+        let objects_installed = self.history.iter().any(|cmd| cmd.installs_objects());
```

Comment:
Good point!

Reply:
I left `installs_objects` because the method is also used to check if we need to update the "paused" status when a new command is sent to a cluster without replicas. But I converted the implementation to an exhaustive match to ensure we won't forget to update it when we introduce new commands.